newIDE/app/src/InstancesEditor/FullSizeInstancesEditorWithScrollbars.js

```diff
@@ -33,154 +79,394 @@ type Props = {|
-    { leading: false, trailing: true }
+    [forceUpdate, hideScrollbarsAfterDelayDebounced]
   );
 
-  render() {
-    const { wrappedEditorRef, ...otherProps } = this.props;
-
-    return (
-      <ScreenTypeMeasurer>
-        {screenType => (
-          <FullSizeMeasurer>
-            {({ width, height }) => (
-              <div style={styles.container}>
-                <InstancesEditor
-                  onViewPositionChanged={
-                    screenType !== 'touch'
-                      ? this._handleViewPositionChange
-                      : noop
-                  }
-                  ref={(editorRef: ?InstancesEditor) => {
-                    wrappedEditorRef && wrappedEditorRef(editorRef);
-                    this._editor = editorRef;
+  const handleViewPositionChange = React.useCallback(
+    (viewPosition: ?ViewPosition) =>
+      throttle(
+        () => {
+          if (!viewPosition) return;
+
+          setAndAdjust({
+            newXValue: viewPosition.getViewX(),
+            newYValue: viewPosition.getViewY(),
+            newWidth: viewPosition.getWidth(),
+            newHeight: viewPosition.getHeight(),
+          });
+        },
+        THROTTLE_TIME, // Throttle the updates after a scroll to avoid make lots of updates in a row that would kill CPU.
+        { leading: false, trailing: true }
+      )(),
+    [setAndAdjust]
+  );
+
+  // Ensure the X Scrollbar doesn't go out of bounds.
+  const minXScrollbarLeftPosition = '0%';
+  const maxXScrollbarLeftPosition = `calc(100% - ${SCROLLBAR_SIZE}px - ${SCROLLBAR_TRACK_WIDTH}px)`;
+  const expectedXScrollbarLeftPosition = `calc(${((xValue.current -
+    xMin.current) /
+    (xMax.current - xMin.current)) *
+    100 +
+    '%'} - ${SCROLLBAR_SIZE}px / 2)`;
+  const xScrollbarLeftPosition = `max(min(${expectedXScrollbarLeftPosition}, ${maxXScrollbarLeftPosition}), ${minXScrollbarLeftPosition})`;
+
+  // Ensure the Y Scrollbar doesn't go out of bounds.
+  const minYScrollbarTopPosition = '0%';
+  const maxYScrollbarTopPosition = `calc(100% - ${SCROLLBAR_SIZE}px - ${SCROLLBAR_TRACK_WIDTH}px)`;
+  const expectedYScrollbarTopPosition = `calc(${((yValue.current -
+    yMin.current) /
+    (yMax.current - yMin.current)) *
+    100 +
+    '%'} - ${SCROLLBAR_SIZE}px / 2)`;
+  const yScrollbarTopPosition = `max(min(${expectedYScrollbarTopPosition}, ${maxYScrollbarTopPosition}), ${minYScrollbarTopPosition})`;
+
+  const screenType = useScreenType();
+
+  return (
+    <FullSizeMeasurer>
+      {({ width, height }) => (
+        <div style={styles.container}>
+          <InstancesEditor
+            onViewPositionChanged={
+              screenType !== 'touch' ? handleViewPositionChange : noop
+            }
+            ref={(editor: ?InstancesEditor) => {
+              wrappedEditorRef && wrappedEditorRef(editor);
+              if (editor) {
+                editorRef.current = editor;
```

Comment:
Isn't this risking to keep an outdated editorRef.current, because of the if (editor) that will prevent editorRef.current to be reset to null? Seem like an edge case an unlikely to happen, but I prefer to understand if there is a real reason to do that - better to have a null error later rather than faking the fact that everything is alright by preventing null to be set.

Reply:
I haven't changed this part of the code.
I don't think the result differs tbh, as editorRef.current is null to start with, setting it or not doesn't change the result?
I'm not sure why it was done like this. I'll update